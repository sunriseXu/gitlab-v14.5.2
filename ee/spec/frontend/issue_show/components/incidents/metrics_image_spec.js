import { GlLink, GlModal } from '@gitlab/ui';
import { createLocalVue, shallowMount, mount } from '@vue/test-utils';
import merge from 'lodash/merge';
import Vuex from 'vuex';
import MetricsImage from 'ee/issue_show/components/incidents/metrics_image.vue';
import createStore from 'ee/issue_show/components/incidents/store';
import waitForPromises from 'helpers/wait_for_promises';

const defaultProps = {
  id: 1,
  filePath: 'test_file_path',
  filename: 'test_file_name',
};

const mockEvent = { preventDefault: jest.fn() };

const localVue = createLocalVue();
localVue.use(Vuex);

describe('Metrics upload item', () => {
  let wrapper;
  let store;

  const mountComponent = (options = {}, mountMethod = mount) => {
    store = createStore();

    wrapper = mountMethod(
      MetricsImage,
      merge(
        {
          localVue,
          store,
          propsData: {
            ...defaultProps,
          },
          provide: { canUpdate: true },
        },
        options,
      ),
    );
  };

  afterEach(() => {
    if (wrapper) {
      wrapper.destroy();
      wrapper = null;
    }
  });

  const findImageLink = () => wrapper.find(GlLink);
  const findCollapseButton = () => wrapper.find('[data-testid="collapse-button"]');
  const findMetricImageBody = () => wrapper.find('[data-testid="metric-image-body"]');
  const findModal = () => wrapper.find(GlModal);
  const findDeleteButton = () => wrapper.find('[data-testid="delete-button"]');

  const closeModal = () => findModal().vm.$emit('hidden');
  const submitModal = () => findModal().vm.$emit('primary', mockEvent);
  const deleteImage = () => findDeleteButton().vm.$emit('click');

  it('render the metrics image component', () => {
    mountComponent({}, shallowMount);

    expect(wrapper.element).toMatchSnapshot();
  });

  it('shows a link with the correct url', () => {
    const testUrl = 'test_url';
    mountComponent({ propsData: { url: testUrl } });

    expect(findImageLink().attributes('href')).toBe(testUrl);
    expect(findImageLink().text()).toBe(defaultProps.filename);
  });

  describe('expand and collapse', () => {
    beforeEach(() => {
      mountComponent();
    });

    it('the card is expanded by default', () => {
      expect(findMetricImageBody().isVisible()).toBe(true);
    });

    it('the card is collapsed when clicked', async () => {
      findCollapseButton().trigger('click');

      await waitForPromises();

      expect(findMetricImageBody().isVisible()).toBe(false);
    });
  });

  describe('delete functionality', () => {
    it('should open the modal when clicked', async () => {
      mountComponent({ stubs: { GlModal: true } });

      deleteImage();

      await waitForPromises();

      expect(findModal().attributes('visible')).toBe('true');
    });

    describe('when the modal is open', () => {
      beforeEach(() => {
        mountComponent(
          {
            data() {
              return { modalVisible: true };
            },
          },
          shallowMount,
        );
      });

      it('should close the modal when cancelled', async () => {
        closeModal();

        await waitForPromises();

        expect(findModal().attributes('visible')).toBeFalsy();
      });

      it('should delete the image when selected', async () => {
        const dispatchSpy = jest.spyOn(store, 'dispatch').mockImplementation(jest.fn());

        submitModal();

        await waitForPromises();

        expect(dispatchSpy).toHaveBeenCalledWith('deleteImage', defaultProps.id);
      });
    });

    describe('canUpdate permission', () => {
      it('delete button is hidden when user lacks update permissions', () => {
        mountComponent({ provide: { canUpdate: false } });

        expect(findDeleteButton().exists()).toBe(false);
      });
    });
  });
});
