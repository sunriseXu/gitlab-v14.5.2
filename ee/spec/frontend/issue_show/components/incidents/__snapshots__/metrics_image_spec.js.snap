// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Metrics upload item render the metrics image component 1`] = `
<gl-card-stub
  bodyclass="gl-border-1,gl-border-t-solid,gl-border-gray-100,[object Object]"
  class="collapsible-card border gl-p-0 gl-mb-5"
  footerclass=""
  headerclass="gl-display-flex gl-align-items-center gl-border-b-0 gl-py-3"
>
  <gl-modal-stub
    actioncancel="[object Object]"
    actionprimary="[object Object]"
    body-class="gl-pb-0! gl-min-h-6!"
    dismisslabel="Close"
    modalclass=""
    modalid="delete-metric-modal"
    size="sm"
    titletag="h4"
  >
     
    <p>
      Are you sure you wish to delete this image?
    </p>
  </gl-modal-stub>
    
  <div
    class="gl-display-flex gl-flex-direction-column"
    data-testid="metric-image-body"
  >
    <img
      class="gl-max-w-full gl-align-self-center"
      src="test_file_path"
    />
  </div>
</gl-card-stub>
`;
